### ISSUE TITLE
ScopeAuthorizeAttribute Does Not Validate Roles or Users

### ISSUE THREAD

**Issue by JoshDobbin** [IdentityModel/Thinktecture.IdentityModel#150]
I may have not been using this attribute correctly. I expected the ScopeAuthorizeAttribute to validate not only the scope but also the roles and users specified in the attribute arguments.

A call to base.IsAuthorized could probably handle some of the other authorization but in my case I was trying to avoid the default redirect behavior that occurs when a user is not authorized (missing a role) but is authenticated.

I modified the ScopeAuthorizeAttribute code to handle these cases in the way I wanted (a flag could be added to enable the default AuthorizeAttribute behavior). I'd be glad to create a pull request but first wanted to make sure I was properly using the attribute.

Here's my modified ScopeAuthorizeAttribute class.

    public class ScopeAuthorizeAttribute : AuthorizeAttribute
    {
        string[] _scopes;
        static string _scopeClaimType = "scope";
        static string _roleClaimType = "role";
        static string _usernameClaimType = "preferred_username";

        public static string ScopeClaimType
        {
            get { return _scopeClaimType; }
            set { _scopeClaimType = value; }
        }

        public ScopeAuthorizeAttribute(params string[] scopes)
        {
            if (scopes == null)
            {
                throw new ArgumentNullException("scopes");
            }

            _scopes = scopes;
        }

        protected override bool IsAuthorized(HttpActionContext actionContext)
        {
            var principal = actionContext.ControllerContext.RequestContext.Principal as ClaimsPrincipal;

            if (principal == null)
            {
                return false;
            }

            return PrincipalIsGrantedAllScopes(principal)
                && PrincipalIsGrantedAllRoles(principal)
                && PrincipalIsAnAllowedUser(principal);
        }

        public bool PrincipalIsGrantedAllScopes(ClaimsPrincipal principal)
        {
            var grantedScopes = principal.FindAll(_scopeClaimType).Select(c => c.Value).ToList();
            foreach (var scope in _scopes)
            {
                if (!grantedScopes.Contains(scope, StringComparer.OrdinalIgnoreCase))
                {
                    return false;
                }
            }

            return true;
        }

        public bool PrincipalIsGrantedAllRoles(ClaimsPrincipal principal)
        {
            var _roles = Roles.Split(new string[] { " " }, StringSplitOptions.RemoveEmptyEntries);
            var grantedRoles = principal.FindAll(_roleClaimType).Select(c => c.Value).ToList();
            foreach (var role in _roles)
            {
                if (!grantedRoles.Contains(role, StringComparer.OrdinalIgnoreCase))
                {
                    return false;
                }
            }

            return true;
        }

        public bool PrincipalIsAnAllowedUser(ClaimsPrincipal principal)

        {
            var _users = Users.Split(new string[] { " " }, StringSplitOptions.RemoveEmptyEntries);
            var usernames = principal.FindAll(_usernameClaimType).Select(c => c.Value).ToList();

            foreach (var user in _users)
            {
                if (usernames.Contains(user, StringComparer.OrdinalIgnoreCase))
                {
                    return true;
                }
            }

            return false;
        }

        protected override void HandleUnauthorizedRequest(HttpActionContext actionContext)
        {
            HttpResponseMessage response;
            var principal = actionContext.ControllerContext.RequestContext.Principal as ClaimsPrincipal;

            if (actionContext.RequestContext.Principal.Identity.IsAuthenticated 
                && !PrincipalIsGrantedAllScopes(principal))
            {
                response = actionContext.Request.CreateErrorResponse(HttpStatusCode.Forbidden, "insufficient_scope");
                response.Headers.Add("WWW-Authenticate", "Bearer error=\"insufficient_scope\"");
            }
            else if (actionContext.RequestContext.Principal.Identity.IsAuthenticated 
                && (!PrincipalIsGrantedAllRoles(principal) || !PrincipalIsAnAllowedUser(principal)))
            {
                response = actionContext.Request.CreateErrorResponse(HttpStatusCode.Forbidden, "insufficient_access");
                response.Headers.Add("WWW-Authenticate", "Bearer error=\"insufficient_access\"");
            }
            else
            {
                response = actionContext.Request.CreateErrorResponse(HttpStatusCode.Unauthorized, "Unauthorized");
            }

            actionContext.Response = response;
        }
    }

**Comment by brockallen** [comment]
Scopes validate that the client has access. This is unrelated to the user having access.

**Comment by JoshDobbin** [comment]
Completely agree. In order to make that more clear, would it be feasible to not have ScopeAuthorize inherit from AuthorizeAttribute? Intellisense in visual studio brings up Roles and Users as optional parameters and they can be set in the attribute arguments. The expectation for me, until I found the implementation, was to expect Roles and Users to be validated as well.

Either way,  I found a solution for me (though I might rename it). Amazing work on all of this. Thank you so much for sharing it.
